### ISSUE TITLE
Idea for better wine support

### ISSUE THREAD

**Issue by basxto** [MikeMaximus/gbm#148]
This is more of a brain storming, but I think there could be some redundancy avoided.
Kinda related to #133

* Use Windows game list for detection with wine
* Extract `$WINEPREFIX` environment variable from process
* Read windows variables from wine registry
* Detect user name  (no idea how)

Environment variables are definitely accessible, since you can manually read them with `ps e <pid>`

Examples from [Freedom Force vs. the 3rd Reich (8890)](https://store.steampowered.com/app/8890/Freedom_Force_vs_the_Third_Reich/) with [Proton](https://github.com/ValveSoftware/Proton):
`WINEPREFIX=/mnt/olhdd/steam/steamapps/compatdata/8890/pfx/`

* `*publicdocs*` => `"Common Documents"="C:\\users\\Public\\Documents"` from $WINEPREFIX/system.reg key `[Software\\Wow6432Node\\Microsoft\\Windows\\CurrentVersion\\Explorer\\Shell Folders]`
* `*mydocs*` => `"Personal"="C:\\users\\steamuser\\My Documents"` from from \*
* `*appdatalocal*` => `"Local AppData"="C:\\users\\steamuser\\Local Settings\\Application Data"` from \*
* `*appdataroaming*` => `"AppData"="C:\\users\\steamuser\\Application Data"` from \*
* `*currentuser*`??

\*) from $WINEPREFIX/userdef.reg or $WINEPREFIX/user.reg key `[Software\\Microsoft\\Windows\\CurrentVersion\\Explorer\\Shell Folders]`

**Comment by MikeMaximus** [comment]
This could work, easier than handling dosbox anyway.  When it's a game running in wine and an absolute save path, I could grab the path variables from user.reg and use those in the path manager.

**Comment by MikeMaximus** [comment]
Also did Valve change the wine binary names with Proton?  Since GBM only looks for wine-preloader or wine64-preloader.  I can't really test it since all my Linux installs are currently VMs.

**Comment by basxto** [comment]
This would be again be paths, GBM can't detect before the game was run the first time.
Therefore it should be similarly handled to #135

**Comment by basxto** [comment]
Best list for proton games so far: https://spcr.netlify.com/

Some games would run best with lutris, because they have complex installers which apply patches and download additional libraries with winetricks.
https://lutris.net/games/one-piece-pirate-warriors-3/ (You appear to have that game too; gamepad fix is wrong)

I haven’t played much with steam controller configuration yet, but that looks like a good way to fix games which have trouble to detect the gamepad correctly. Especially since you can download and share them.

There are still many games with problems, but it would be naive to expect ports to run better. https://wiki.archlinux.org/index.php/Steam/Game-specific_troubleshooting

SteamOS + Steam Controller will benefit heavily from proton. Same goes for Steam Link (device and app)

**Comment by basxto** [comment]
Maybe prefixes for MonitorID should be introduced in general.
Otherwise you would need to dublicate monitorlist completely.
`L` for Linux, `W` for Windows / Wine.
Others could follow like `M` for Mac, `B` for BSD, `G` for GNU, `H` for Haiku, `A` for AmigaOS/AROS etc.
But I don’t think alternative open source operating systems have much to offer in regard to gaming. They likely “only“ have ported open source games and run Windows and Linux games through compatibility layers.

**Comment by MikeMaximus** [comment]
Yeah i'm still chewing on how to handle some parts of this issue.  

The Wine functionality part is basically done and needs some testing / tweaking.  I'm just not sure how I want to handle a few issues in the GUI and the configuration identity issue you discussed.

**Comment by MikeMaximus** [comment]
This is how this feature will work for v1.1.5.

1.  In Linux, users are now able to import from either the official Windows or Linux configuration lists.  A one-time information message about this feature is the displayed the first time a Linux user tries to import from the Windows list.

2.  Windows configurations are displayed in the Game Manager as they are in Windows, except environment variables won't be resolved.

3.  The first time a Windows configuration with a convert-able save path is detected running in Wine/Proton, GBM will convert the save path for the configuration to what it would be within the prefix and save it. 

I decided not to make any additions for configuration OS identify for this version.

I need to do some testing with Wine games that use a relative path before release,  but the feature seems to work great so far with all the Proton games I've tried that use an absolute path.

**Comment by basxto** [comment]
Sounds good, thanks.
So in game manager you won’t see whether a game is native or not?

**Comment by MikeMaximus** [comment]
> So in game manager you won’t see whether a game is native or not?

Right, there's no special flag yet to designate which OS a configuration is designed for.  I'm just using tags to keep things organized at the moment.

**Comment by basxto** [comment]
Maybe just add tag “Wine” to games downloaded from the windows list. Unless you are doing something similar already.

**Comment by MikeMaximus** [comment]
I was thinking about that,  as well as dropping the official tag from them since they'll be modified once the game is detected.

**Comment by basxto** [comment]
Why do they get modified?
So on linux you can’t export to the official list of Windows games? It would be interesting adding games, one tested with Wine.

**Comment by MikeMaximus** [comment]
The save path is modified to point to the wine prefix, so it's ready for future saves and so you're able to restore if required.  

It would be nice to have an amalgamated game list, it's just not something I want to tackle for this release.

**Comment by MikeMaximus** [comment]
> Somehow misread that as game path. So you are avoiding to create a new field for the prefix.

Yeah basically.  I have some changes in mind that I want to do all that once, but i'd like to get this release out first.

**Comment by MikeMaximus** [comment]
I've made the following changes when importing from the Windows list in Linux.

1.  Wine tag is automatically added.
2.  Official tag is removed.
3.  Games tagged with DOSBox / ScummVM are not shown for import.

**Comment by basxto** [comment]
With known `prefix` path and path of the `emulator` executable it be done much easier, I think. (related to #151)

```sh
#!/bin/sh
wine='wine'
export WINEPREFIX="${HOME}/.wine"
export COMMONDOCUMENTS=$(${wine} reg query 'HKLM\Software\Wow6432Node\Microsoft\Windows\CurrentVersion\Explorer\Shell Folders' /v 'Common Documents' | egrep -o ".?:\\\\.*" | tr -d '\r')
export USERDOCUMENTS=$(${wine} reg query 'HKCU\Software\Microsoft\Windows\CurrentVersion\Explorer\Shell Folders' /v 'Personal' | egrep -o ".?:\\\\.*" | tr -d '\r')
resolve(){
  resolved=$(${wine} cmd /c echo $1)
  ${wine} winepath "${resolved}"
}

resolve '%COMMONDOCUMENTS%\Monolith Productions\Condemned\Save'
resolve '%LOCALAPPDATA%Low\Weather Factory\Cultist Simulator'
resolve '%PROGRAMDATA%\Steam\Bejeweled2\users'
resolve '%APPDATA%\DarkSoulsII'
resolve '%USERDOCUMENTS%\nbgi\darksouls'
resolve '%USERPROFILE%\Saved Games\darksiders1.exe'
```

```
/home/dabascht/.wine/dosdevices/c:/users/Public/Documents/Monolith Productions/Condemned/Save                                                                 
/home/dabascht/.wine/dosdevices/c:/users/dabascht/Local Settings/Application DataLow/Weather Factory/Cultist Simulator                                        
/home/dabascht/.wine/dosdevices/c:/users/Public/Application Data/Steam/Bejeweled2/users                                                                       
/home/dabascht/.wine/dosdevices/c:/users/dabascht/Application Data/DarkSoulsII
/home/dabascht/.wine/dosdevices/c:/users/dabascht/Meine Dokumente/nbgi/darksouls                                                                              
/home/dabascht/.wine/dosdevices/c:/users/dabascht/Saved Games/darksiders1.exe
```

**Comment by MikeMaximus** [comment]
That is a more future proof way to do things.  GBM's current method will break if they change how Wine reg files are stored/formated.
